infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -249,42 +269,98 @@ private void updateReplicationParameters(Set<LogReplicationSession> sessionsUnch
+                    return null;
+                } catch (TransactionAbortedException e) {
+                    log.error("Failed to create sessions.  Retrying.", e);
+                    throw new RetryNeededException();
+                }
+            }).run();
+        } catch (InterruptedException e) {
+            log.error("Unrecoverable Corfu Error when creating the sessions", e);
+            throw new UnrecoverableCorfuInterruptedError(e);
+        }
 
+        log.info("Total of {} outgoing sessions created with subscriber {}, sessions={}", sessionsToAdd.size(),
+                subscriber, sessionsToAdd);
+
+        configManager.generateConfig(sessionsToAdd);
+    }
+
+    private void createIncomingSessionsBySubscriber(ReplicationSubscriber subscriber) {
+        Set<LogReplicationSession> sessionsToAdd = new HashSet<>();
+        try {
+            String localClusterId = topology.getLocalClusterDescriptor().getClusterId();
+            IRetry.build(IntervalRetry.class, () -> {
+                try (TxnContext txn = corfuStore.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                    // Create out-going sessions by combing the given registered replication subscriber and topology
                     for(ClusterDescriptor remoteSourceCluster : topology.getRemoteSourceClusters().values()) {
-                        LogReplicationSession session = constructSession(remoteSourceCluster.getClusterId(), localClusterId);
-                        if (!sessions.contains(session)) {
-                            newSessionsDiscovered.add(session);
-                            incomingSessionsToAdd.add(session);
-                            metadataManager.addSession(txn, session, topology.getTopologyConfigId(), true);
+                        Set<ReplicationModel> supportedModels =
+                                topology.getRemoteSourceClusterToReplicationModels().get(remoteSourceCluster);
+                        if (supportedModels.contains(subscriber.getModel())) {
+                            LogReplicationSession session =
+                                    constructSession(remoteSourceCluster.clusterId, localClusterId, subscriber);
+                            // TODO: (V2 / Chris) this is still not thread-safe, need to synchronize on sessions
+                            if (!sessions.contains(session)) {
+                                sessionsToAdd.add(session);
+                                newSessionsDiscovered.add(session);
+                                metadataManager.addSession(txn, session, topology.getTopologyConfigId(), true);
+                            }
                         }
                     }
-
                     txn.commit();
+                    sessions.addAll(sessionsToAdd);
+                    incomingSessions.addAll(sessionsToAdd);
                     return null;
                 } catch (TransactionAbortedException e) {
```

Comment:
Don't we need to clear the sessionsToAdd before retrying - otherwise it will carry stale data from previous retry right?

Reply:
Yes good catch! Moved the class fields updating logic to be out of the retry, and cleared the method's `sessionsToAdd` map before retrying.